File: Problems/1498. (M) Number of Subsequences That Satisfy the Given Sum Condition/Java/solution.java
class Solution {
  public int numSubseq(int[] nums, int target) {
    // sort with two pointers

    Arrays.sort(nums);
    int l = 0;
    int h = nums.length - 1;
    int mod = 1_000_000_007;
    int ret = 0;

    // precompute powers of 2
    int[] pow = new int[nums.length];
    pow[0] = 1;
    for (int i = 1; i < nums.length; i++)
      pow[i] = (pow[i - 1] * 2) % mod;

    while (l <= h) {
      if (nums[l] + nums[h] > target)
        h--;
      else {
        ret += pow[h-l];
        ret %= mod;
        l++;
      }
    }
    return ret;
  }
}

File: Problems/1498. (M) Number of Subsequences That Satisfy the Given Sum Condition/Python3/solution.py
class Solution:
  def numSubseq(self, nums: List[int], target: int) -> int:
    # sort with two pointers

    nums.sort()
    l, h = 0, len(nums)-1
    ret = 0
    mod = 10**9 + 7
    while l <= h:
      if nums[l] + nums[h] > target:
        h -= 1
      else:
        ret += 2**(h-l)
        ret %= mod
        l += 1

    return ret

